### PR TITLE
SLVS-890 [Technical] Improve the performance of S2933 (field should b…

### PR DESCRIPTION
**Pull request by tvsonar**
…e readonly)

### CODE REVIEW

File: src/SonarLint.CSharp/Rules/FieldShouldBeReadonly.cs [hunk 1]
```diff
@@ -58,97 +59,71 @@ public class FieldShouldBeReadonly : SonarDiagnosticAnalyzer
...
 
-                analysisContext.RegisterCompilationEndAction(c =>
-                {
-                    var fields = candidateFields.Except(nonCandidateFields);
-                    fields = fields.Intersect(assignedAsReadonly);
+                    var fields = assignedAsReadonly.Intersect(candidateFields).Except(nonCandidateFields);
+
                     foreach (var field in fields)
                     {
                         var declarationReference = field.DeclaringSyntaxReferences.FirstOrDefault();
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/FieldShouldBeReadonly.cs]: the syntax could be collected beforehand, we are only reporting in the current file.

File: src/SonarLint.CSharp/Rules/FieldShouldBeReadonly.cs [hunk 2]
```diff
@@ -158,16 +133,91 @@ protected override void Initialize(SonarAnalysisContext context)
...
-            });
+                },
+                SyntaxKind.ClassDeclaration,
+                SyntaxKind.InterfaceDeclaration);
+        }
+
+        private static void CollectFieldsFromPartial(SyntaxNodeSemanticModelTuple<TypeDeclarationSyntax> partialDeclaration,
+            HashSet<IFieldSymbol> assignedAsReadonly, HashSet<IFieldSymbol> nonCandidateFields)
+        {
+            var assignments = partialDeclaration.SyntaxNode.DescendantNodes()
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/FieldShouldBeReadonly.cs]: split these into separate methods.

File: src/SonarLint.CSharp/Rules/FieldShouldBeReadonly.cs [hunk 3]
```diff
@@ -158,16 +133,91 @@ protected override void Initialize(SonarAnalysisContext context)
...
+                .OfType<AssignmentExpressionSyntax>()
+                .Where(a => assignmentKinds.Contains(a.Kind()));
+
+            foreach (var assignment in assignments)
+            {
+                ProcessExpressionChange(assignment.Left, partialDeclaration.SemanticModel, nonCandidateFields, assignedAsReadonly);
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/FieldShouldBeReadonly.cs]: should be named `CollectFieldFrom...`

**Comment by tvsonar** [comment]
There's false positive on the following:
```
class X
{
    private int x;
    public X()
    {
        new X().x = 12;
    }
}
```

This is not new, but should be fixed. (We need to check `this.x`, or simply `x`, any other reference can't be made `readonly`)

**Comment by dbolkensteyn** [comment]
LGTM

**Pull request merged**